### PR TITLE
ananicy-cpp: support wrapped applications

### PR DESCRIPTION
**Pull request by diniamo**
## Description of changes

Ananicy-cpp now matches wrapped applications with this patch.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by diniamo** [comment]
I can't find how to request reviews so: @JohnRTitor @Artturin

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ 
+ #include <chrono>
+@@ -29,7 +30,10 @@ void Worker::work(const std::stop_token &stop_token) {
+   while (!stop_token.stop_requested()) {
+     while ((proc = process_queue->poll(500ms)).has_value()) {
+       const auto &p = proc.value();
+-      const auto &rule = rules->get_rule(p.name);
++      auto name = p.name;
++      if (utils::regex::is_str_matching_regex(name, "^\\..+-wrap(p(ed?)?)?$"))
++        name = name.substr(1, name.find_last_of('-') - 1);
```
Comment (Artturin) [inline pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: So it matches -wrapped and then removes the -wrapped right?

What's up with the `(p(ed?)?)`?

File: pkgs/by-name/an/ananicy-cpp/package.nix [hunk 2]
```diff
@@ -33,7 +34,8 @@ clangStdenv.mkDerivation rec {
       url = "https://gitlab.com/ananicy-cpp/ananicy-cpp/-/commit/6ea2dccceec39b6c4913f617dad81d859aa20f24.patch";
       hash = "sha256-C+7x/VpVwewXEPwibi7GxGfjuhDkhcjTyGbZHlYL2Bs=";
     })
-  ];
+  ]
+  ++ lib.optional wrapperSupport ./match-wrappers.patch;
```
Comment (Artturin) [inline pkgs/by-name/an/ananicy-cpp/package.nix:38]: Let's just always add it

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch (hunk 1 above)
Comment (JohnRTitor) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: AFAIK, sometimes process names don't fully have "-wrapped" suffix in their name, due to size limitations.

File: pkgs/by-name/an/ananicy-cpp/package.nix (hunk 2 above)
Comment (JohnRTitor) [reply pkgs/by-name/an/ananicy-cpp/package.nix:38]: Agreed, is this on upstream patch, by the way? If not, I think it should go to upstream directly.

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch (hunk 1 above)
Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: It matches
```
.x-wrap
.x-wrapp
.x-wrappe
.x-wrapped
```
and removes everything after the last `-` (the `-` included).

File: pkgs/by-name/an/ananicy-cpp/package.nix (hunk 2 above)
Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/package.nix:38]: No, I wrote it.

Comment (Artturin) [reply pkgs/by-name/an/ananicy-cpp/package.nix]: > Agreed, is this an upstream patch, by the way? If not, I think it should go to upstream directly.

It's nix specific so idk if upstream would accept it, though we can try.

Comment (JohnRTitor) [reply pkgs/by-name/an/ananicy-cpp/package.nix]: Not all process names end like that, as you can see. 

https://github.com/CachyOS/ananicy-rules/blob/e3638ebb2320202ede9facfc9657cc8af2098a2d/00-default/Services/gvfs.rules#L10

I had to manually add these rules to match `.x-` `.x-w`, `.x-wra` and so on.

All wrapped app's processes have `.` in the beginning. 
I think the rules should run on any process `x` and `.x+.`

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch (hunk 1 above)
Comment (JohnRTitor) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: Not all process names end like that, as you can see. 

https://github.com/CachyOS/ananicy-rules/blob/e3638ebb2320202ede9facfc9657cc8af2098a2d/00-default/Services/gvfs.rules

image


I had to manually add these rules to match `.x-` `.x-w`, `.x-wra` and so on.

All wrapped app's processes have `.` in the beginning. 
I think the rules should run on any process `x` and `\.x.*` (regex)

Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: @JohnRTitor are you sure that screenshots includes the full process names?

Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: Looks like they do, in which case I'll try to come up with a better way of stripping the extra characters. I'm not sure how yet, so ideas are welcome.

Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:20]: Never mind, it looks like `ps` is doing something funky. If you check the [nixpkgs makeWrapper implementation](https://github.com/NixOS/nixpkgs/blob/683aa7c4e385509ca651d49eeb35e58c7a1baad6/pkgs/build-support/setup-hooks/make-wrapper.sh#L219), you can see that wrapers always start with a `.` and end with `-wrapped`. You can also confirm this by check the store paths for programs, eg.: `ls -a $(dirname $(realpath $(which firefox)))`. And fortunately the process names in ananicy-cpp are also correct, why I confirmed by logging:

tmp ElXKWq8NkR

(The right side is `ps`, the left side is a log line from ananicy-cpp)

**Comment by diniamo** [comment]
@JohnRTitor @Artturin could we finish this up?

**Comment by VeilSilence** [comment]
Is there anything need to do to merge this?

**Comment by JohnRTitor** [comment]
> @JohnRTitor @Artturin could we finish this up?

Looks good, let me do a quick check before I approve.

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
++++ b/src/worker.cpp
+@@ -29,7 +29,10 @@ void Worker::work(const std::stop_token &stop_token) {
+   while (!stop_token.stop_requested()) {
+     while ((proc = process_queue->poll(500ms)).has_value()) {
+       const auto &p = proc.value();
+-      const auto &rule = rules->get_rule(p.name);
++      auto name = p.name;
++      if (name.starts_with('.') && name.ends_with("-wrapped"))
++        name = name.substr(1, name.find_last_of('-') - 1);
++      const auto &rule = rules->get_rule(name);
```
Comment (Artturin) [inline pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:13]: Does the `.` have to be removed?

**Comment by JohnRTitor** [comment]
By the way, is there a reason we are not using `ananicy-cpp` as the default in `services.ananicy`?

**Review by JohnRTitor — APPROVED** [review]
LGTM based on my testing, approved

File: pkgs/by-name/an/ananicy-cpp/match-wrappers.patch (hunk 3 above)
Comment (diniamo) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:13]: Yes, it does. Both the `.` and the `-wrapped` are appended by the wrapper script, so we want the part between them for matching.

Comment (Artturin) [reply pkgs/by-name/an/ananicy-cpp/match-wrappers.patch:13]: I meant is the . being removed, I guess the (1 is doing that? Not familiar with cpp

**Comment by diniamo** [comment]
@JohnRTitor can you make a pr to remove all the now reduant rules in the cachyos rules repo?

**Comment by JohnRTitor** [comment]
I think those can stay, for those using just `ananicy` (non cpp) and who are not on the version with the patch. 

Future rules to the repo, won't need to account for wrapped processes.

**Comment by Pandapip1** [comment]
FYI, having not seen the discussion on this PR, I opened https://github.com/CachyOS/ananicy-rules/pull/174 and it was merged

**Comment by diniamo** [comment]
I'm still fine with that, and at this point, John is probably as well. This patch is in the latest stable release.

**Pull request merged**